File: sdks/node/src/plugins/droid-policy.ts
import fsp from 'node:fs/promises';
import os from 'node:os';
import path from 'node:path';
import { inspectPhysicalDirectory, readPhysicalFile, type FileSnapshot } from './managed-files.js';
import { parseStrictJsoncObject, type JsonObject } from './strict-json.js';

const MAX_SOURCE_BYTES = 2 * 1024 * 1024;

interface PolicyLocation {
  readonly filePath: string;
  readonly label: string;
}

interface PolicyLayer extends PolicyLocation {
  readonly snapshot?: FileSnapshot;
  readonly hooksDisabled?: boolean;
  readonly allowManagedHooksOnly?: boolean;
  readonly showHookOutput?: boolean;
}

export interface DroidPolicyPrecondition {
  readonly filePath: string;
  readonly label: string;
  readonly maximumBytes: number;
  readonly original?: FileSnapshot;
}

export interface DroidPolicyOrigin {
  readonly filePath: string;
  readonly label: string;
}

export interface DroidPolicyState {
  readonly allowManagedHooksOnlyBy?: DroidPolicyOrigin;
  readonly hooksDisabled?: boolean;
  readonly hooksDisabledBy?: DroidPolicyOrigin;
  readonly preconditions: readonly DroidPolicyPrecondition[];
}

function hasCode(error: unknown, code: string): boolean {
  return typeof error === 'object' && error !== null && 'code' in error && error.code === code;
}

function optionalBoolean(root: JsonObject, field: string, label: string): boolean | undefined {
  const value = root[field];
  if (value !== undefined && typeof value !== 'boolean') {
    throw new Error(`${label} field "${field}" must be a boolean`);
  }
  return value as boolean | undefined;
}

function managedSettingsPath(): string {
  if (process.platform === 'darwin') {
    return '/Library/Application Support/Factory/settings.json';
  }
  if (process.platform === 'win32') {
    return path.join(process.env.ProgramFiles ?? 'C:\\Program Files', 'Factory', 'settings.json');
  }
  return '/etc/factory/settings.json';
}

async function readLayer(location: PolicyLocation): Promise<PolicyLayer> {
  await inspectPhysicalDirectory(path.dirname(location.filePath), `${location.label} directory`);
  const snapshot = await readPhysicalFile(location.filePath, location.label, MAX_SOURCE_BYTES);
  if (!snapshot) return location;
  const root = parseStrictJsoncObject(snapshot.contents, `${location.label} at ${location.filePath}`);
  return {
    ...location,
    snapshot,
    hooksDisabled: optionalBoolean(root, 'hooksDisabled', location.label),
    allowManagedHooksOnly: optionalBoolean(root, 'allowManagedHooksOnly', location.label),
    showHookOutput: optionalBoolean(root, 'showHookOutput', location.label),
  };
}

async function gitRoot(start: string): Promise<string | undefined> {
  let current = path.resolve(start);
  for (;;) {
    const marker = path.join(current, '.git');
    try {
      const metadata = await fsp.lstat(marker);
      if (metadata.isSymbolicLink() || (!metadata.isDirectory() && !metadata.isFile())) {
        throw new Error(`Factory Droid project marker is not physical: ${marker}`);
      }
      return current;
    } catch (error) {
      if (!hasCode(error, 'ENOENT') && !hasCode(error, 'ENOTDIR')) throw error;
    }
    const parent = path.dirname(current);
    if (parent === current) return undefined;
    current = parent;
  }
}

function projectDirectories(root: string, current: string): string[] {
  const directories = [root];
  const relative = path.relative(root, current);
  if (!relative || relative.startsWith('..') || path.isAbsolute(relative)) return directories;
  let directory = root;
  for (const segment of relative.split(path.sep).filter(Boolean)) {
    directory = path.join(directory, segment);
    directories.push(directory);
  }
  return directories;
}

async function projectLocations(): Promise<PolicyLocation[][]> {
  const current = path.resolve(process.cwd());
  const root = await gitRoot(current) ?? current;
  return projectDirectories(root, current).map((directory, index) => {
    const factory = path.join(directory, '.factory');
    const scope = index === 0 ? 'project' : `folder ${directory}`;
    return [{
      filePath: path.join(factory, 'settings.json'),
      label: `Factory Droid ${scope} settings`,
    }, {
      filePath: path.join(factory, 'settings.local.json'),
      label: `Factory Droid ${scope} local settings`,
    }];
  });
}

function scopeValue(layers: readonly PolicyLayer[], field: 'hooksDisabled'): PolicyLayer | undefined {
  const [settings, local] = layers;
  return local[field] !== undefined ? local : settings[field] !== undefined ? settings : undefined;
}

export async function readDroidPolicy(): Promise<DroidPolicyState> {
  const managedLocation: PolicyLocation = {
    filePath: managedSettingsPath(),
    label: 'Factory Droid system-managed settings',
  };
  const scopes = await projectLocations();
  const [managed, ...projectLayers] = await Promise.all([
    readLayer(managedLocation),
    ...scopes.map(async (scope) => Promise.all(scope.map(readLayer))),
  ]);
  const flattened = [managed, ...projectLayers.flat()];
  const allowManagedHooksOnlyBy = managed.allowManagedHooksOnly === true
    ? { filePath: managed.filePath, label: managed.label }
    : undefined;
  const selected = [managed, ...projectLayers.map((scope) => scopeValue(scope, 'hooksDisabled'))]
    .find((layer) => layer?.hooksDisabled !== undefined);
  return {
    allowManagedHooksOnlyBy,
    hooksDisabled: selected?.hooksDisabled,
    hooksDisabledBy: selected
      ? { filePath: selected.filePath, label: selected.label }
      : undefined,
    preconditions: flattened.map((layer) => ({
      filePath: layer.filePath,
      label: layer.label,
      maximumBytes: MAX_SOURCE_BYTES,
      original: layer.snapshot,
    })),
  };
}

File: sdks/go/cmd/elydora/plugins/droid_policy.go
package plugins

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
)

type droidPolicyLocation struct {
	filePath string
	label    string
}

type droidPolicyLayer struct {
	filePath              string
	label                 string
	snapshot              *managedFileSnapshot
	hooksDisabled         *bool
	allowManagedHooksOnly *bool
	showHookOutput        *bool
}

type droidPolicyOrigin struct {
	filePath string
	label    string
}

type droidPolicyState struct {
	allowManagedHooksOnlyBy *droidPolicyOrigin
	hooksDisabled           *bool
	hooksDisabledBy         *droidPolicyOrigin
	preconditions           []filePrecondition
}

func defaultDroidManagedSettingsPath() string {
	switch runtime.GOOS {
	case "darwin":
		return "/Library/Application Support/Factory/settings.json"
	case "windows":
		programFiles := os.Getenv("ProgramFiles")
		if programFiles == "" {
			programFiles = `C:\Program Files`
		}
		return filepath.Join(programFiles, "Factory", "settings.json")
	default:
		return "/etc/factory/settings.json"
	}
}

var droidManagedSettingsPath = defaultDroidManagedSettingsPath

func readDroidPolicyLayer(location droidPolicyLocation) (*droidPolicyLayer, error) {
	if _, err := managedPhysicalDirectoryExists(
		filepath.Dir(location.filePath),
		location.label+" directory",
	); err != nil {
		return nil, err
	}
	snapshot, err := readManagedFile(
		location.filePath,
		location.label,
		maxManagedSourceBytes,
	)
	if err != nil {
		return nil, err
	}
	layer := &droidPolicyLayer{
		filePath: location.filePath,
		label:    location.label,
		snapshot: snapshot,
	}
	if snapshot == nil {
		return layer, nil
	}
	root, err := decodeJSONCObject(
		snapshot.contents,
		fmt.Sprintf("%s at %s", location.label, location.filePath),
		true,
	)
	if err != nil {
		return nil, err
	}
	layer.hooksDisabled, err = droidOptionalBoolean(root, "hooksDisabled", location.label)
	if err != nil {
		return nil, err
	}
	layer.allowManagedHooksOnly, err = droidOptionalBoolean(
		root,
		"allowManagedHooksOnly",
		location.label,
	)
	if err != nil {
		return nil, err
	}
	layer.showHookOutput, err = droidOptionalBoolean(root, "showHookOutput", location.label)
	if err != nil {
		return nil, err
	}
	return layer, nil
}

func droidGitRoot(start string) (string, bool, error) {
	current, err := filepath.Abs(start)
	if err != nil {
		return "", false, fmt.Errorf("resolve Factory Droid project directory: %w", err)
	}
	for {
		marker := filepath.Join(current, ".git")
		info, inspectErr := os.Lstat(marker)
		switch {
		case inspectErr == nil:
			if info.Mode()&os.ModeSymlink != 0 || (!info.IsDir() && !info.Mode().IsRegular()) {
				return "", false, fmt.Errorf(
					"Factory Droid project marker is not physical: %s",
					marker,
				)
			}
			return current, true, nil
		case !os.IsNotExist(inspectErr):
			return "", false, fmt.Errorf(
				"inspect Factory Droid project marker at %s: %w",
				marker,
				inspectErr,
			)
		}
		parent := filepath.Dir(current)
		if parent == current {
			return "", false, nil
		}
		current = parent
	}
}

func droidProjectDirectories(root, current string) []string {
	directories := []string{root}
	relative, err := filepath.Rel(root, current)
	if err != nil || relative == "." || relative == "" || filepath.IsAbs(relative) {
		return directories
	}
	if relative == ".." || len(relative) > 3 && relative[:3] == ".."+string(os.PathSeparator) {
		return directories
	}
	directory := root
	for _, segment := range splitDroidPath(relative) {
		directory = filepath.Join(directory, segment)
		directories = append(directories, directory)
	}
	return directories
}

func splitDroidPath(value string) []string {
	parts := make([]string, 0)
	for value != "." && value != "" {
		directory, file := filepath.Split(value)
		if file != "" {
			parts = append([]string{file}, parts...)
		}
		value = filepath.Clean(directory)
		if value == string(os.PathSeparator) {
			break
		}
	}
	return parts
}

func droidProjectPolicyLocations() ([][2]droidPolicyLocation, error) {
	current, err := os.Getwd()
	if err != nil {
		return nil, fmt.Errorf("resolve Factory Droid working directory: %w", err)
	}
	current, err = filepath.Abs(current)
	if err != nil {
		return nil, fmt.Errorf("resolve Factory Droid working directory: %w", err)
	}
	root, found, err := droidGitRoot(current)
	if err != nil {
		return nil, err
	}
	if !found {
		root = current
	}
	directories := droidProjectDirectories(root, current)
	locations := make([][2]droidPolicyLocation, 0, len(directories))
	for index, directory := range directories {
		scope := "project"
		if index > 0 {
			scope = "folder " + directory
		}
		factory := filepath.Join(directory, ".factory")
		locations = append(locations, [2]droidPolicyLocation{
			{filepath.Join(factory, "settings.json"), "Factory Droid " + scope + " settings"},
			{filepath.Join(factory, "settings.local.json"), "Factory Droid " + scope + " local settings"},
		})
	}
	return locations, nil
}

func droidScopeHooksDisabled(
	settings, local *droidPolicyLayer,
) *droidPolicyLayer {
	if local.hooksDisabled != nil {
		return local
	}
	if settings.hooksDisabled != nil {
		return settings
	}
	return nil
}

func readDroidPolicy() (*droidPolicyState, error) {
	managed, err := readDroidPolicyLayer(droidPolicyLocation{
		droidManagedSettingsPath(),
		"Factory Droid system-managed settings",
	})
	if err != nil {
		return nil, err
	}
	locations, err := droidProjectPolicyLocations()
	if err != nil {
		return nil, err
	}
	projectLayers := make([][2]*droidPolicyLayer, 0, len(locations))
	for _, scope := range locations {
		settings, readErr := readDroidPolicyLayer(scope[0])
		if readErr != nil {
			return nil, readErr
		}
		local, readErr := readDroidPolicyLayer(scope[1])
		if readErr != nil {
			return nil, readErr
		}
		projectLayers = append(projectLayers, [2]*droidPolicyLayer{settings, local})
	}
	state := &droidPolicyState{}
	if managed.allowManagedHooksOnly != nil && *managed.allowManagedHooksOnly {
		state.allowManagedHooksOnlyBy = &droidPolicyOrigin{managed.filePath, managed.label}
	}
	selected := managed
	if selected.hooksDisabled == nil {
		selected = nil
		for _, scope := range projectLayers {
			if layer := droidScopeHooksDisabled(scope[0], scope[1]); layer != nil {
				selected = layer
				break
			}
		}
	}
	if selected != nil {
		state.hooksDisabled = selected.hooksDisabled
		state.hooksDisabledBy = &droidPolicyOrigin{selected.filePath, selected.label}
	}
	layers := []*droidPolicyLayer{managed}
	for _, scope := range projectLayers {
		layers = append(layers, scope[0], scope[1])
	}
	state.preconditions = make([]filePrecondition, 0, len(layers))
	for _, layer := range layers {
		state.preconditions = append(state.preconditions, filePrecondition{
			filePath:    layer.filePath,
			label:       layer.label,
			snapshot:    layer.snapshot,
			maximumSize: maxManagedSourceBytes,
		})
	}
	return state, nil
}
